modules/packages/ParallelIO.chpl

```diff
@@ -0,0 +1,705 @@
+ * limitations under the License.
+ */
+
+/*
+  Helper procedures for doing parallel I/O
+
+  This module provides the following procedures, each intended for a slightly
+  different use case (but all focused on cases where a large file containing
+  only a header, and then a continuous stream of deserializeable values of the
+  same type).
+
+  There are distributed and local versions of each procedure. The distributed
+  procedures return a block distributed array over the provided locales, while
+  the local procedures return a default rectangular array.
+
+  * :proc:`readParallelLines`: read each of the lines of a file as a ``string``
+    or ``bytes`` value
+  * :proc:`readParallelDelimited`: read a file where each value is strictly
+    separated by a delimiter, and the delimiter cannot be found in the
+    value (e.g., CSV)
+  * :proc:`readParallel`: read a file where each value is separated by delimiter
+
+  There are also some helper procedures that can be used to implement other
+  parallel I/O procedures:
+
+  * :proc:`findDelimChunks`: find a set of byte offsets that divide a file into
+    roughly equal chunks where each chunk begins with a delimiter
+  * :proc:`findDelimChunksChecked`: find a set of byte offsets that divide a file
+    into roughly equal chunks where each chunk begins with a delimiter and each
+    chunk starts with a deserializable value
+  * :proc:`findItemOffsets`: get a prefix sum of the number of items in each
+    chunk of a file, where the chunks are defined by the ``byteOffsets`` array,
+    and each item is separated by the given delimiter
+*/
+@unstable("the 'parallelIO' module is unstable and subject to change in a future release")
+module ParallelIO {
+  private use IO, BlockDist, List, CTypes, OS;
+
+  /*
+    Read a file's lines in parallel into a block-distributed array.
+
+    This routine is similar to :proc:`readParallelDelimited`, except that it
+    reads each line as a :type:`~String.string` or :type:`~Bytes.bytes` value.
+
+    :arg filePath: a path to the file to read from
+    :arg lineType: which type to represent a line: either ``string`` or ``bytes``
+    :arg nTasks: the number of tasks to use per locale
+        (if ``-1``, query ``here.maxTaskPar`` on each locale)
+    :arg header: how to handle the file header (see :record:`headerPolicy`)
+    :arg targetLocales: the locales to read the file on
+
+    :returns: a block-distributed array of ``lineType`` values
+
+    :throws: :class:`OffsetNotFoundError` if a starting offset cannot be found
+             in any of the chunks
+
+    See :proc:`~IO.open` for other errors that could be thrown when attempting
+  */
+  proc readParallelLines(filePath: string, type lineType = string, header = headerPolicy.noHeader,
+                         nTasks: int = -1, targetLocales: [?d] locale = Locales
+  ): [] lineType throws
+    where lineType == string || lineType == bytes
+  {
```

Comment:
Why doesn't this just call `readParallelDelimited` and pass `b"\n"` as the delimiter?

Reply:
Because passing an itemType of `string` or `bytes` to `readParallelDelimited` would cause it to read the entire file into a string as a single item.

Reply:
It's not obvious to me why it would do that. Do you think there is some other way that these functions could be combined, though?

Reply:
I think with the default deserializer, this line would keep reading until the end of the file when `item` is a `string`
https://github.com/jeremiah-corrado/chapel/blob/abff59e5b11b0bba21dd98254be81d67cf2d3fc5/modules/packages/ParallelIO.chpl#L245
Is that correct?

They could potentially be combined if the "kernel" (the part that actually reads the items) can be passed into a more general procedure as an fcf, since that's the only place where the two procedures differ.

Reply:
Default deserializer for strings/bytes reads one word.

Reply:
Okay, so I don't think `readParallelDelimited("path/to/file.txt", t=string, delim="\n")` would always behave the same as r`eadParallelLines("path/to/file.txt")`, because there could be multiple strings between each delimiter due to whitespace. Note that the docs for readParallelDelimited proc also indicate that each `t` must be separated by exactly one delimiter.